pkg/monitoring/monitoring.go

```diff
@@ -310,38 +310,38 @@ func (d *derivedFloat64Metric) Name() string {
 	return d.name
 }
 
-// no-op
 func (d *derivedFloat64Metric) Register() error {
 	return nil
 }
 
 func (d *derivedFloat64Metric) ValueFrom(valueFn func() float64, labelValues ...string) {
-	if len(labelValues) == 0 {
-		if err := d.base.UpsertEntry(valueFn); err != nil {
-			log.Errorf("failed to add value for derived metric %q: %v", d.name, err)
-		}
-		return
-	}
-	lv := make([]metricdata.LabelValue, 0, len(labelValues))
-	for _, l := range labelValues {
-		lv = append(lv, metricdata.NewLabelValue(l))
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	if len(labelValues) != len(d.keys) {
+		panic(fmt.Errorf("metric %v has %v keys but provided %v values", d.name, len(d.keys), len(labelValues)))
 	}
-	if err := d.base.UpsertEntry(valueFn, lv...); err != nil {
-		log.Errorf("failed to add value for derived metric %q: %v", d.name, err)
+	kv := make([]attribute.KeyValue, 0, len(d.keys))
+	for i, k := range d.keys {
+		v := labelValues[i]
+		kv = append(kv, attribute.KeyValue(k.Value(v)))
 	}
+	as := attribute.NewSet(kv...)
+	d.attrs[as] = valueFn
 }
 
 type float64Metric struct {
-	*stats.Float64Measure
+	mt  MeterType
+	c   api.Float64Counter
```

Comment:
this looks odd, maybe should review the design of Metric interface?

Reply:
The alternative is to have 3 concrete implementations. Which we could do but I don't think this is so bad?

I guess we could have 4 structs - a  common one with common logic and 3 for (counter,guage,distribution)? I can see if that is cleaner.

BTW this is all somewhat hacky, I was planning to clean it up if we move forward

Reply:
> The alternative is to have 3 concrete implementations.

I try this way, it makes PR too huge.